check.ml

```diff
@@ -0,0 +1,118 @@
+open Instr
+
+exception MissingMain
+exception DuplicateFunctionDeclaration of identifier
+exception InvalidFunctionDeclaration of identifier
+exception DuplicateVersion of identifier * label
+exception EmptyFunction of identifier
+exception DuplicateParameter of identifier * variable
+
+exception ErrorAt of identifier * label * exn
+
+(* Those should always come wrapped in an ErrorAt : *)
+exception FunctionDoesNotExist of identifier
+exception VersionDoesNotExist of identifier * label
+exception InvalidNumArgs of pc
+exception InvalidArgument of pc * expression
+exception MissingReturn
+
+
+module IdentifierSet = Set.Make(Identifier)
+
+let well_formed prog =
+  (* Check if main exists and expects no arguments *)
+  let check_main main =
+    if main.name <> "main" then raise MissingMain;
+    if main.formals <> [] then raise (InvalidFunctionDeclaration "main");
+  in
+  check_main prog.main;
+
+  let lookup_version func label =
+    try List.find (fun {label=l} -> label = l) func.body with
+    | Not_found -> raise (VersionDoesNotExist (func.name, label))
+  in
+  let lookup_fun name =
+    if name = "main" then prog.main else
+    try List.find (fun {name = l} -> name = l) prog.functions with
+    | Not_found -> raise (FunctionDoesNotExist name)
+  in
+
+  let functions = prog.main :: prog.functions in
+
+  (* Formals args shall not contain duplicate variables *)
+  let check_formals name formals =
+    let formals = List.map (fun f -> match f with
+        | ParamConst x -> x | ParamMut x -> x) formals in
+    List.iter (fun f ->
+        if List.length (List.filter (fun f' -> f=f') formals) > 1 then
+          raise (DuplicateParameter (name, f))) formals
+  in
+
+  let check_version func version =
+    let instrs = version.instrs in
+
+    if func.name <> "main" then
+    begin if Array.length instrs = 0 then raise MissingReturn else
```

Comment:
I liked @gasche's [earlier comment](https://github.com/reactorlabs/sourir/pull/77#discussion_r109983318) about using `;` instead of `else`. It's a lot like writing asserts/ifs-without-else at the beginning of C functions.